### DISCUSSION TITLE
Using .ifref to conditionally include library routines

### DISCUSSION THREAD

**Discussion by Ster0pes** [cc65/cc65#1396] [Q&A] [answered]
Hello!

At the moment I'm working on a simple game written in assembly for the Commander X16.
I'm trying to create a general object file (vera.o) containing code pertaining to the VERA chip. 
I'd like to be able to only include the code for the routines from vera.o that I'm actually using in the game. If I'm not gravely mistaking, .ifref should help me with that, given that the documentation says this about it: "This command may be used to build subroutine libraries in include files (you may use separate object modules for this purpose too)."

So what I did is this:
vera.inc:
.global clearScreen

vera.asm:
.include "vera.inc"
...
.ifref clearScreen
.proc clearScreen
...
.endproc
.endif

;game.asm
.include "vera.inc"
...
jsr clearScreen

Assembling my game yields this error:
Unresolved external 'clearScreen' referenced in:
  program.asm(12)
ld65: Error: 1 unresolved external(s) found - cannot create output file

It works when I remove the .ifref, so that seems to be the problem..,
What am I doing wrong here? Is it possible at all to conditionally include library subroutines using cc65? If so, how?

**Comment by groessler** [comment, marked as answer]
The proper way to do this is to put each function in a separate object 
file and create a library file out of them.

Then link with the library file and the linker does the magic.

**Reply by Ster0pes** [reply]
Sorry, I'm still very new to this. What would the solution you're suggesting look like?

**Reply by groessler** [reply]
The tool you need is `ar65` (https://cc65.github.io/doc/ar65.html). Libraries are called "archives" in this documentation since they are nothing more than archives of object files.

The linker links complete object files to the executable. So the idea is that an object file should only contain what's needed for its functionality. In general, every function of your lib in its own object file.
You can put more functions in one object file, if they ever only can be used when all of them are in the executable (they reference each other).

So, the default way: Assume the clearScreen function is written in a way that it writes space chars (using a putChar function) until the screen is clear. Then clearScreen should be in its own file, referencing an external putChar function. putChar in turn should also be in its own file.
Now, if the main program just uses putChar to print something, the code for clearScreen is left out of the executable.

Hope I was clear...

**Reply by Ster0pes** [reply]
I believe so, thanks for pointing me in the right direction. This is how I got it working:
* Create testlib.inc, containing:
   .global testMethod
   .global anotherMethod`
* Create testMethod.asm, in which I include testlib.inc and implement testMethod
* Create anotherMethod.asm, in which I also include testlib.inc and implement anotherMethod
* Assemble testMethod.asm to testMethod.o
* Assemble anotherMethod.asm to anotherMethod.o
* Use ar65 to add testMethod.o and anotherMethod.o to testlib.lib
* Include testlib.inc in my main program and call testMethod and/or anotherMethod
* Assemble my main program including testlib.lib in the call to cl65
Is this the proper way to do it?

Still, out of curiosity:
This explanation under .ifref in the documentation:
"This command may be used to build subroutine libraries in include files (you may use separate object modules for this purpose too)."
What did the writer mean by this? He does mention the object module approach, but what about the .ifref thing? In what situation(s) can that be used the way it's described here?

**Reply by groessler** [reply]
> Is this the proper way to do it?

Yes


Regarding `.ifref`. It doesn't do forward references, the symbol needs to have been referenced before the `.ifref` line. This is also written in the documentation: "The condition is true if if the the given symbol was referenced before, ..."

Example:
```
$ cat mylib.inc 
; mylib.inc  -- my library
.ifref clearScreen
clearScreen:
        nop     ; blabla code
        jsr putChar
        nop
        rts
.endif
.ifref putChar
putChar:
        sta $1234
        rts
.endif
$ cat main1.s   
; main1.s  -  first test program
.code
.include "mylib.inc"
.export start
start:
        lda #' '
        jmp putChar
.end
$ cat main2.s 
; main2.s  -  second test program
.code
.export start
start:
        lda #' '
        jmp putChar
.include "mylib.inc"
.end
$ cat main3.s 
; main3.s  -  third test program
.code
.export start
start:
        jmp clearScreen
.include "mylib.inc"
.end
```

Compile the three test programs:

```
$ cl65 -t atari -C atari-asm.cfg -Wa -lmain1.lst main1.s -o main1.com
main1.s(7): Error: Symbol 'putChar' is undefined
$ cl65 -t atari -C atari-asm.cfg -Wa -lmain2.lst main2.s -o main2.com
$ cl65 -t atari -C atari-asm.cfg -Wa -lmain3.lst main3.s -o main3.com
```

`main1.s` didn't work because the `.ifref` was before the symbol use.

Listing files:

```
$ cat main2.lst 
ca65 V2.19 - Git 855143123
Main file   : main2.s
Current file: main2.s

000000r 1               ; main2.s  -  second test program
000000r 1               .code
000000r 1               .export start
000000r 1               start:
000000r 1  A9 20                lda #' '
000002r 1  4C rr rr             jmp putChar
000005r 1               .include "mylib.inc"
000005r 2               ; mylib.inc  -- my library
000005r 2               .ifref clearScreen
000005r 2               clearScreen:
000005r 2                       nop     ; blabla code
000005r 2                       jsr putChar
000005r 2                       nop
000005r 2                       rts
000005r 2               .endif
000005r 2               .ifref putChar
000005r 2               putChar:
000005r 2  8D 34 12             sta $1234
000008r 2  60                   rts
000009r 2               .endif
000009r 2               
000009r 1               .end
$ cat main3.lst 
ca65 V2.19 - Git 855143123
Main file   : main3.s
Current file: main3.s

000000r 1               ; main3.s  -  third test program
000000r 1               .code
000000r 1               .export start
000000r 1               start:
000000r 1  4C rr rr             jmp clearScreen
000003r 1               .include "mylib.inc"
000003r 2               ; mylib.inc  -- my library
000003r 2               .ifref clearScreen
000003r 2               clearScreen:
000003r 2  EA                   nop     ; blabla code
000004r 2  20 rr rr             jsr putChar
000007r 2  EA                   nop
000008r 2  60                   rts
000009r 2               .endif
000009r 2               .ifref putChar
000009r 2               putChar:
000009r 2  8D 34 12             sta $1234
00000Cr 2  60                   rts
00000Dr 2               .endif
00000Dr 2               
00000Dr 1               .end
```

**Reply by oliverschmidt** [reply]
@groessler: Thanks for explaining both the library and the `.ifref` approach :-)

**Reply by Ster0pes** [reply]
@groessler Thanks for your elaborate answer. I think I understand now what I did wrong. I had mylib.inc (containing .global clearScreen) and mylib.asm (containing the implementation of the method), and I only included myLib.inc in my main program. Oops.

Good thing I messed up though and ended up asking a question here. Otherwise I wouldn't have learned about ar65, which (as you said) is a way more elegant solution.

**Reply by oliverschmidt** [reply]
@Ster0pes: Thanks for the nice feedback :-)
